### PR TITLE
CAN Support

### PR DESCRIPTION
**Pull request by azaleacolburn**
### Description
Send, receive, and use CAN device data between a robot project and fission.

### Objectives
- [x] Send TalonFX Data
- [x] Send CANSparkMax Data
- [x] Make simulated motor fields appear as CAN devices rather than SimDevices
- [x] CAN Motor Brain Mappings
- [x] CANEncoder support

### Testing Done
- [x] Test TalonFX in WS Viewer
- [x] Test CANSparkMax in WS Viewer
- [x] Make sure CAN Motor inputs map into the brain properly
- [x] Test CANEncoders on SparkMaxes

For testing:
1. Build and publish SyntheSim with `./gradlew build` and `./gradlew publishToMavenLocal`.
2. Build and run the JavaSample with `./gradlew build` and `./gradlew simulateJava`.
3. Run fission
4. Open fission, spawn a robot, then click the WS Test button in the debug menu (this must be done first).
5. Verify the connection by opening the WS Viewer panel, you should see a long list of motors and encoders.
6. Click the RoboRio button to setup CAN output group, select all the CANSparkMax motors and all the WheelDrivers, then click Done
7. Do the same but for Encoders, this time, just select CANSparkMax[1] and any of the wheel stimuli.
8. In the robotSimulation screen, click Autonomous, you should see the robot drive forward 20 (feet?), then stop.

We need to have a discussion about how we want device mapping to work, right now every motor in a group's output is averaged and applied to each driver. We need to make it easier to do one to one mappings between motors and drivers, because right now, clicking through the whole menu ~10 times to make an output group for each motor is ridiculous.

> [!note]
> Pulled from #1077 for it's encoder framework, so I think just turning that PR into a gyro / joystick PR is the way to go.

> [!note]
> CANMotor inputs (eg. supplyCurrent) have been ignored, since they're not simulated in fission.

This PR is a combination of JIRA Issues [1747](https://jira.autodesk.com/browse/AARD-1747), [1748](https://jira.autodesk.com/browse/AARD-1748), and [1749](https://jira.autodesk.com/browse/AARD-1749) for a total of 18 story points.

### CODE REVIEW

**Review by BrandonPacewic — COMMENTED** [review]
~~Managed to get everything built but had some difficulty running the auto and teleop successfully. I couldn't even get the previous testing I did for #1043 to work. Is there some specific order you need to run each component in?~~

File: fission/src/systems/simulation/wpilib_brain/SimInput.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    private _stimulus: EncoderStimulus
+
+    constructor(device: string, stimulus: EncoderStimulus) {
+        this._device = device
+        this._stimulus = stimulus
+    }
+
+    public Update(_deltaT: number) {
+        SimCANEncoder.SetPosition(`${this._device}`, this._stimulus.positionValue)
+        SimCANEncoder.SetVelocity(`${this._device}`, this._stimulus.velocityValue)
```
Comment (BrandonPacewic) [inline fission/src/systems/simulation/wpilib_brain/SimInput.ts]: Is there a reason for using template literals here when `_device` is already a string?

File: fission/src/systems/simulation/wpilib_brain/SimInput.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    private GetZ(): number {
+        return this.GetAxis(SimGyroInput.AXIS_Z)
+    }
+
+    public Update(_deltaT: number) {
+        const x = this.GetX()
+        const y = this.GetY()
+        const z = this.GetZ()
+        // console.log(`${this._device}\n${x}\n${y}\n${z}`)
```
Comment (BrandonPacewic) [inline fission/src/systems/simulation/wpilib_brain/SimInput.ts]:
```suggestion
```

Still needed?

File: fission/src/systems/simulation/wpilib_brain/SimOutput.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        })
+    }
+}
+
+export class CANOutputGroup extends SimOutputGroup {
+    public constructor(name: string, ports: number[], drivers: Driver[]) {
+        super(name, ports, drivers, SimType.CANMotor)
+    }
+
+    // Averaging is probably not the right solution
```
Comment (BrandonPacewic) [inline fission/src/systems/simulation/wpilib_brain/SimOutput.ts]: Probably best to update this comment and describe the solution we might switch to in the future. Not everyone will remember this from internal discussions.

File: fission/src/systems/simulation/wpilib_brain/WPILibBrain.ts [hunk 4]
```diff
@@ -141,62 +155,83 @@ export class SimCAN {
...
 
-    public static GetDutyCycle(device: string): number | undefined {
-        return SimGeneric.Get("CANMotor", device, CANMOTOR_DUTY_CYCLE, 0.0)
+    // public static GetDutyCycle(device: string): number | undefined {
+    //     return SimGeneric.Get("CANMotor", device, CANMOTOR_DUTY_CYCLE, 0.0)
+    // }
+    //
+    // public static SetSupplyVoltage(device: string, voltage: number): boolean {
+    //     return SimGeneric.Set("CANMotor", device, CANMOTOR_SUPPLY_VOLTAGE, voltage)
+    // }
```
Comment (BrandonPacewic) [inline fission/src/systems/simulation/wpilib_brain/WPILibBrain.ts]: Still needed?

File: fission/vite.config.ts [hunk 5]
```diff
@@ -1,76 +1,85 @@
...
 // https://vitejs.dev/config/
 export default defineConfig({
-    plugins: [react(), /* viteSingleFile() */ glsl({
-    include: [                   // Glob pattern, or array of glob patterns to import
-      '**/*.glsl', '**/*.wgsl',
-      '**/*.vert', '**/*.frag',
-      '**/*.vs', '**/*.fs'
+    plugins: [
+        react(),
+        /* viteSingleFile() */ glsl({
```
Comment (BrandonPacewic) [inline fission/vite.config.ts]: I see that the formatter caught this file. (Which is totally awesome!) But we should probably remove the `viteSingleFile()` plugin comment as a part of this formatting change.

File: simulation/SyntheSimJava/src/main/java/com/autodesk/synthesis/ctre/TalonFXConfigurator.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package com.autodesk.synthesis.ctre;
+
+import com.ctre.phoenix6.hardware.DeviceIdentifier;
+import com.ctre.phoenix6.configs.TorqueCurrentConfigs;
+import com.ctre.phoenix6.StatusCode;
+
+//import com.ctre.phoenix6.configs.TalonFXConfigurator
```
Comment (BrandonPacewic) [inline simulation/SyntheSimJava/src/main/java/com/autodesk/synthesis/ctre/TalonFXConfigurator.java]: Leaving this for the future?

File: simulation/SyntheSimJava/src/main/java/com/autodesk/synthesis/revrobotics/CANSparkMax.java [hunk 7]
```diff
@@ -1,7 +1,11 @@
 package com.autodesk.synthesis.revrobotics;
 
+import java.util.ArrayList;
+
 import com.autodesk.synthesis.CANEncoder;
 import com.autodesk.synthesis.CANMotor;
+// import com.autodesk.synthesis.revrobotics.SparkAbsoluteEncoder;
```
Comment (BrandonPacewic) [inline simulation/SyntheSimJava/src/main/java/com/autodesk/synthesis/revrobotics/CANSparkMax.java]: Same as previous, leaving this for the future?

File: simulation/samples/JavaSample/src/main/java/frc/robot/Robot.java [hunk 8]
```diff
@@ -15,117 +12,164 @@
...
+    private String m_autoSelected;
+    private final SendableChooser<String> m_chooser = new SendableChooser<>();
+
+    private CANSparkMax m_SparkMax1 = new CANSparkMax(1, MotorType.kBrushless);
+    private CANSparkMax m_SparkMax2 = new CANSparkMax(2, MotorType.kBrushless);
+    private CANSparkMax m_SparkMax3 = new CANSparkMax(3, MotorType.kBrushless);
+    private CANSparkMax m_SparkMax4 = new CANSparkMax(4, MotorType.kBrushless);
+    private CANSparkMax m_SparkMax5 = new CANSparkMax(5, MotorType.kBrushless);
+    private CANSparkMax m_SparkMax6 = new CANSparkMax(6, MotorType.kBrushless);
+    //private com.autodesk.synthesis.ctre.TalonFX m_Talon = new com.autodesk.synthesis.ctre.TalonFX(2);
```
Comment (BrandonPacewic) [inline simulation/samples/JavaSample/src/main/java/frc/robot/Robot.java:38]: Leaving this for the future?

File: simulation/samples/JavaSample/src/main/java/frc/robot/Robot.java [hunk 9]
```diff
@@ -15,117 +12,164 @@
...
+     * You can add additional auto modes by adding additional comparisons to the
+     * switch structure
+     * below with additional strings. If using the SendableChooser make sure to add
+     * them to the
+     * chooser code above as well.
+     */
+    @Override
+    public void autonomousInit() {
+        m_autoSelected = m_chooser.getSelected();
+        //m_autoSelected = SmartDashboard.getString("Auto Selector", kDefaultAuto);
```
Comment (BrandonPacewic) [inline simulation/samples/JavaSample/src/main/java/frc/robot/Robot.java:86]: Leaving this for the future?

**Review by BrandonPacewic — COMMENTED** [review]
Ok this is pretty sick. Only very minor nitpicks from me.

> [!note]
> To anyone else reviewing this PR, make sure to hit the `WS Test` button on the debug panel to properly connect the websocket after following the instructions from #1043 BEFORE attempting to make a motor or encoder group.

File: fission/src/systems/simulation/wpilib_brain/SimInput.ts (hunk 1 above)
Comment (azaleacolburn) [reply fission/src/systems/simulation/wpilib_brain/SimInput.ts]: Nope =)
It was incase some other formatting needed to be done later at the time.

File: fission/vite.config.ts (hunk 5 above)
Comment (azaleacolburn) [reply fission/vite.config.ts]: yes! that's so cool!

File: simulation/SyntheSimJava/src/main/java/com/autodesk/synthesis/ctre/TalonFXConfigurator.java (hunk 6 above)
Comment (azaleacolburn) [reply simulation/SyntheSimJava/src/main/java/com/autodesk/synthesis/ctre/TalonFXConfigurator.java]: wdym?

File: simulation/samples/JavaSample/src/main/java/frc/robot/Robot.java (hunk 8 above)
Comment (azaleacolburn) [reply simulation/samples/JavaSample/src/main/java/frc/robot/Robot.java:38]: That can be commented in for testing

File: simulation/samples/JavaSample/src/main/java/frc/robot/Robot.java (hunk 9 above)
Comment (azaleacolburn) [reply simulation/samples/JavaSample/src/main/java/frc/robot/Robot.java:86]: that was some testing @PepperLola was doing

File: simulation/SyntheSimJava/src/main/java/com/autodesk/synthesis/ctre/TalonFXConfigurator.java (hunk 6 above)
Comment (BrandonPacewic) [reply simulation/SyntheSimJava/src/main/java/com/autodesk/synthesis/ctre/TalonFXConfigurator.java]: Wondering if this was also left for testing or not.

Comment (azaleacolburn) [reply simulation/SyntheSimJava/src/main/java/com/autodesk/synthesis/ctre/TalonFXConfigurator.java]: it was not

File: fission/src/systems/simulation/wpilib_brain/WPILibBrain.ts [hunk 10]
```diff
@@ -1,24 +1,35 @@
-/* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (BrandonPacewic) [inline fission/src/systems/simulation/wpilib_brain/WPILibBrain.ts:1]: Not sure if you meant to delete this or not but the linter is certainly not happy about it.

**Review by HunterBarclay — COMMENTED** [review]
I'm unable to build the sample project because it says it can't find CTRE or REV libraries:
image

I built and published SyntheSimJava successfully, however.

**Comment by BrandonPacewic** [comment]
> I'm unable to build the sample project because it says it can't find CTRE or REV libraries
> 
> I built and published SyntheSimJava successfully, however.

@HunterBarclay Did you follow the following instructions from #1043?

> 1. Navigate to simulation/samples/JavaSample
> 2. Install [WPILib](https://docs.wpilib.org/en/stable/docs/zero-to-robot/step-2/wpilib-setup.html) utilities if you don't have them
> 3. Open the JavaSample project in WPILib VSCode
> 4. Install Phoenix v6 and REVLib vendor libraries
>     1. https://maven.ctr-electronics.com/release/com/ctre/phoenix6/latest/Phoenix6-frc2024-latest.json
>     2. https://software-metadata.revrobotics.com/REVLib-2024.json
> 5. Run ./gradlew build && ./gradlew simulateJava

File: fission/src/systems/simulation/wpilib_brain/WPILibBrain.ts (hunk 10 above)
Comment (azaleacolburn) [reply fission/src/systems/simulation/wpilib_brain/WPILibBrain.ts:1]: Fixed

**Comment by HunterBarclay** [comment]
@azaleacolburn Why aren't the vendor deps just already added?

**Comment by azaleacolburn** [comment]
> @azaleacolburn Why aren't the vendor deps just already added?

In every wpilib project I've been on, we've always installed locally. Same reason we don't upload node_modules to github (but obviously to a lesser extent). And that won't matter once we have them install from the autodesk maven. But if we want them to install the library locally, then uploading the vendordeps would probably be a good idea.

**Review by a-crowell — CHANGES_REQUESTED** [review]
This is so cool! And motors are connecting great! Encoder isn't connecting for me though, so auto doesn't stop.

**Review by a-crowell — APPROVED** [review]
Super impressive! If I knew how to get a device from id, I'd let you know, but alas I am without that knowledge and the code is working great. So excited for code sim!

**Review by BrandonPacewic — APPROVED** [review]
Super awesome!

**Pull request merged**